core/primitives/src/offchain.rs

```diff
@@ -0,0 +1,297 @@
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Timestamp(u64);
+
+/// Duration type
+#[derive(Clone, Copy, PartialEq, Eq, Ord, PartialOrd, Default)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Duration(u64);
+
+impl Duration {
+	/// Create new duration representing given number of milliseconds.
+	pub fn from_millis(millis: u64) -> Self {
+		Duration(millis)
+	}
+
+	/// Returns number of milliseconds this Duration represents.
+	pub fn millis(&self) -> u64 {
+		self.0
+	}
+}
+
+impl Timestamp {
+	/// Creates new `Timestamp` given unix timestamp in miliseconds.
+	pub fn from_unix_millis(millis: u64) -> Self {
+		Timestamp(millis)
+	}
+
+	/// Increase the timestamp by given `Duration`.
+	pub fn add(&self, duration: Duration) -> Timestamp {
+		Timestamp(self.0.saturating_add(duration.0))
+	}
+
+	/// Decrease the timestamp by given `Duration`
+	pub fn sub(&self, duration: Duration) -> Timestamp {
+		Timestamp(self.0.saturating_sub(duration.0))
+	}
+
+	/// Returns a saturated difference (Duration) between two Timestamps.
+	pub fn diff(&self, other: &Self) -> Duration {
+		Duration(self.0.saturating_sub(other.0))
+	}
+
+	/// Return number of milliseconds since UNIX epoch.
+	pub fn unix_millis(&self) -> u64 {
+		self.0
+	}
+}
+
+/// An extended externalities for offchain workers.
+pub trait Externalities {
+	/// Submit transaction.
+	///
+	/// The transaction will end up in the pool and be propagated to others.
+	fn submit_transaction(&mut self, extrinsic: Vec<u8>) -> Result<(), ()>;
+
+	/// Sign given piece of data with current authority key.
+	///
+	/// Returns `None` if signing is not available.
+	fn sign(&mut self, data: &[u8]) -> Option<[u8; 64]>;
+
+	/// Returns current UNIX timestamp (in millis)
+	fn timestamp(&mut self) -> Timestamp;
+
+	/// Pause the execution until `deadline` is reached.
+	fn sleep_until(&mut self, deadline: Timestamp);
+
+	/// Returns a random seed.
+	///
+	/// This is a trully random non deterministic seed generated by host environment.
+	/// Obviously fine in the off-chain worker context.
+	fn random_seed(&mut self) -> [u8; 32];
+
+	/// Sets a value in the local storage.
+	///
+	/// Note this storage is not part of the consensus, it's only accessible by
+	/// offchain worker tasks running on the same machine. It IS persisted between runs.
+	fn local_storage_set(&mut self, key: &[u8], value: &[u8]);
+
+	/// Reads a value from the local storage.
+	///
+	/// If the value does not exist in the storage `None` will be returned.
+	/// Note this storage is not part of the consensus, it's only accessible by
+	/// offchain worker tasks running on the same machine. It IS persisted between runs.
+	fn local_storage_read(&mut self, key: &[u8]) -> Option<Vec<u8>>;
+
+	/// Initiaties a http request given HTTP verb and the URL.
+	///
+	/// Meta is a future-reserved field containing additional, parity-codec encoded parameters.
+	/// Returns the id of newly started request.
+	fn http_request_start(
```

Comment:
I wonder what is the lifetime of a request / response. Looks like I can read headers after the request finished. Does they go away after the off-chain worker finishes (they finish at some point, don't they?)?

Reply:
to clarify: can off-chain workers run indefinitely, like in a while loop?